File: Algorithms/Flatten 2D Vector/solution.py
class Vector2D(object):
    #Method 1: Load all elements first
    def __init__(self, vec2d):
        """
        Initialize your data structure here.
        :type vec2d: List[List[int]]
        """
        self.list = [num for nums in vec2d for num in nums]

    def next(self):
        """
        :rtype: int
        """
        return self.list.pop(0) if self.list else None

    def hasNext(self):
        """
        :rtype: bool
        """
        return bool(self.list)

    #Method 2: Load one by one
    def __init__(self, vec2d):
        """
        Initialize your data structure here.
        :type vec2d: List[List[int]]
        """
        self.row, self.col, self.list = 0, 0, vec2d

    def next(self):
        """
        :rtype: int
        """
        val = self.list[self.row][self.col]
        self.col += 1
        return val

    def hasNext(self):
        """
        :rtype: bool
        """
        while self.row < len(self.list):
            if self.col < len(self.list[self.row]):
                return True
            self.row += 1
            self.col = 0
        return False

File: Algorithms/Flatten 2D Vector/solution.java
public class Vector2D {
    Iterator<List<Integer>> row;
    Iterator<Integer> col;
    public Vector2D(List<List<Integer>> vec2d) {
        row = vec2d.iterator();
        col = null;
    }

    public int next() {
        if (!hasNext()) return -1;
        return col.next();
    }

    public boolean hasNext() {
	//Use while instead of if in case of list contains empty list
        while (row.hasNext() && (col == null || !col.hasNext())) col = row.next().iterator();
        //check col != null, in case of the whole list is null/empty
	return col != null && col.hasNext();
    }
}

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D i = new Vector2D(vec2d);
 * while (i.hasNext()) v[f()] = i.next();
 */
